Error using log4j2 api/core libraries with Android

## Description

Using log4j2 v2.24.0 (api and core) with Android throws error:

    I/com.example: Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.management.ManagementFactory" on path: DexPathList[[zip file "/data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/base.apk"],nativeLibraryDirectories=[/data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/lib/x86, /data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/base.apk!/lib/x86, /system/lib, /system_ext/lib]

gradle settings:

    implementation ('org.apache.logging.log4j:log4j-api:2.24.1'){
        exclude group: 'com.sun.jmx', module: 'jmx'
        exclude group: 'javax.management'
    }
    implementation 'org.apache.logging.log4j:log4j-core:2.24.1'

## Configuration

**Version:** 2.24.0

**Operating system:** Android SDK34

**JDK:** 1.8

## Logs
    I/com.example: Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.management.ManagementFactory" on path: DexPathList[[zip file "/data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/base.apk"],nativeLibraryDirectories=[/data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/lib/x86, /data/app/~~MLF2ZshZzBIhJrgJjHw41g==/com.example-3Fbkjvp4obnGh8xK-dY_6A==/base.apk!/lib/x86, /system/lib, /system_ext/lib]]

Full stack trace as follows:

> I/xample.log4jap: Rejecting re-init on previously-failed class java.lang.Class<org.apache.logging.log4j.core.lookup.JmxRuntimeInputArgumentsLookup>: java.lang.NoClassDefFoundError: Failed resolution of: Ljava/lang/management/ManagementFactory;
>         at void org.apache.logging.log4j.core.lookup.JmxRuntimeInputArgumentsLookup.<clinit>() (JmxRuntimeInputArgumentsLookup.java:35)
>         at java.lang.Object java.lang.reflect.Constructor.newInstance0(java.lang.Object[]) (Constructor.java:-2)
>         at java.lang.Object java.lang.reflect.Constructor.newInstance(java.lang.Object[]) (Constructor.java:343)
>         at java.lang.Object org.apache.logging.log4j.core.util.ReflectionUtil.instantiate(java.lang.Class) (ReflectionUtil.java:188)
>         at void org.apache.logging.log4j.core.lookup.Interpolator.<init>(org.apache.logging.log4j.core.lookup.StrLookup, java.util.List) (Interpolator.java:90)
>         at void org.apache.logging.log4j.core.lookup.Interpolator.<init>(java.util.Map) (Interpolator.java:109)
>         at void org.apache.logging.log4j.core.config.AbstractConfiguration.<init>(org.apache.logging.log4j.core.LoggerContext, org.apache.logging.log4j.core.config.ConfigurationSource) (AbstractConfiguration.java:137)
>         at void org.apache.logging.log4j.core.config.NullConfiguration.<init>() (NullConfiguration.java:32)
>         at void org.apache.logging.log4j.core.LoggerContext.<clinit>() (LoggerContext.java:76)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.createContext(java.lang.String, java.net.URI) (ClassLoaderContextSelector.java:265)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.lambda$new$0$org-apache-logging-log4j-core-selector-ClassLoaderContextSelector() (ClassLoaderContextSelector.java:57)
>         at java.lang.Object org.apache.logging.log4j.core.selector.ClassLoaderContextSelector$$ExternalSyntheticLambda1.get() (D8$$SyntheticClass:-1)
>         at java.lang.Object org.apache.logging.log4j.util.LazyUtil$SafeLazy.value() (LazyUtil.java:113)
>         at java.lang.Object org.apache.logging.log4j.util.Lazy.get() (Lazy.java:39)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getDefault() (ClassLoaderContextSelector.java:273)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getContext(java.lang.String, java.lang.ClassLoader, java.util.Map$Entry, boolean, java.net.URI) (ClassLoaderContextSelector.java:152)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getContext(java.lang.String, java.lang.ClassLoader, boolean, java.net.URI) (ClassLoaderContextSelector.java:125)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(java.lang.String, java.lang.ClassLoader, java.lang.Object, boolean, java.net.URI, java.lang.String) (Log4jContextFactory.java:241)
>         at org.apache.logging.log4j.spi.LoggerContext org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(java.lang.String, java.lang.ClassLoader, java.lang.Object, boolean, java.net.URI, java.lang.String) (Log4jContextFactory.java:46)
>         at org.apache.logging.log4j.spi.LoggerContext org.apache.logging.log4j.LogManager.getContext(boolean) (LogManager.java:118)
>         at org.apache.logging.log4j.Logger org.apache.logging.log4j.LogManager.getLogger(java.lang.String) (LogManager.java:622)
>         at org.apache.logging.log4j.Logger org.apache.logging.log4j.LogManager.getRootLogger() (LogManager.java:657)
>         at void com.example.log4japi.MainActivity.<init>() (MainActivity.java:28)
>         at java.lang.Object java.lang.Class.newInstance() (Class.java:-2)
>         at android.app.Activity android.app.AppComponentFactory.instantiateActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (AppComponentFactory.java:95)
> I/xample.log4jap:     at android.app.Activity androidx.core.app.CoreComponentFactory.instantiateActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (CoreComponentFactory.java:45)
>         at android.app.Activity android.app.Instrumentation.newActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (Instrumentation.java:1253)
>         at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:3353)
>         at android.app.Activity android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.app.servertransaction.PendingTransactionActions, android.content.Intent) (ActivityThread.java:3601)
>         at void android.app.servertransaction.LaunchActivityItem.execute(android.app.ClientTransactionHandler, android.os.IBinder, android.app.servertransaction.PendingTransactionActions) (LaunchActivityItem.java:85)
>         at void android.app.servertransaction.TransactionExecutor.executeCallbacks(android.app.servertransaction.ClientTransaction) (TransactionExecutor.java:135)
>         at void android.app.servertransaction.TransactionExecutor.execute(android.app.servertransaction.ClientTransaction) (TransactionExecutor.java:95)
>         at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:2066)
>         at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
>         at void android.os.Looper.loop() (Looper.java:223)
>         at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:7656)
>         at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
>         at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:592)
>         at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:947)
>      Caused by: java.lang.ClassNotFoundException: Didn't find class "java.lang.management.ManagementFactory" on path: DexPathList[[dex file "/data/data/com.example.log4japi/code_cache/.overlay/base.apk/classes.dex", dex file "/data/data/com.example.log4japi/code_cache/.overlay/base.apk/classes3.dex", dex file "/data/data/com.example.log4japi/code_cache/.overlay/base.apk/classes4.dex", zip file "/data/app/~~IM0jLBxKndj1Iv1NsWAUrA==/com.example.log4japi-gluwbHKsoguOXmqwA790hA==/base.apk"],nativeLibraryDirectories=[/data/app/~~IM0jLBxKndj1Iv1NsWAUrA==/com.example.log4japi-gluwbHKsoguOXmqwA790hA==/lib/x86, /system/lib, /system_ext/lib]]
>         at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:207)
>         at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:379)
>         at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
>         at void org.apache.logging.log4j.core.lookup.JmxRuntimeInputArgumentsLookup.<clinit>() (JmxRuntimeInputArgumentsLookup.java:35)
>         at java.lang.Object java.lang.reflect.Constructor.newInstance0(java.lang.Object[]) (Constructor.java:-2)
>         at java.lang.Object java.lang.reflect.Constructor.newInstance(java.lang.Object[]) (Constructor.java:343)
>         at java.lang.Object org.apache.logging.log4j.core.util.ReflectionUtil.instantiate(java.lang.Class) (ReflectionUtil.java:188)
>         at void org.apache.logging.log4j.core.lookup.Interpolator.<init>(org.apache.logging.log4j.core.lookup.StrLookup, java.util.List) (Interpolator.java:90)
>         at void org.apache.logging.log4j.core.lookup.Interpolator.<init>(java.util.Map) (Interpolator.java:109)
>         at void org.apache.logging.log4j.core.config.AbstractConfiguration.<init>(org.apache.logging.log4j.core.LoggerContext, org.apache.logging.log4j.core.config.ConfigurationSource) (AbstractConfiguration.java:137)
>         at void org.apache.logging.log4j.core.config.NullConfiguration.<init>() (NullConfiguration.java:32)
>         at void org.apache.logging.log4j.core.LoggerContext.<clinit>() (LoggerContext.java:76)
> I/xample.log4jap:     at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.createContext(java.lang.String, java.net.URI) (ClassLoaderContextSelector.java:265)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.lambda$new$0$org-apache-logging-log4j-core-selector-ClassLoaderContextSelector() (ClassLoaderContextSelector.java:57)
>         at java.lang.Object org.apache.logging.log4j.core.selector.ClassLoaderContextSelector$$ExternalSyntheticLambda1.get() (D8$$SyntheticClass:-1)
>         at java.lang.Object org.apache.logging.log4j.util.LazyUtil$SafeLazy.value() (LazyUtil.java:113)
>         at java.lang.Object org.apache.logging.log4j.util.Lazy.get() (Lazy.java:39)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getDefault() (ClassLoaderContextSelector.java:273)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getContext(java.lang.String, java.lang.ClassLoader, java.util.Map$Entry, boolean, java.net.URI) (ClassLoaderContextSelector.java:152)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.selector.ClassLoaderContextSelector.getContext(java.lang.String, java.lang.ClassLoader, boolean, java.net.URI) (ClassLoaderContextSelector.java:125)
>         at org.apache.logging.log4j.core.LoggerContext org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(java.lang.String, java.lang.ClassLoader, java.lang.Object, boolean, java.net.URI, java.lang.String) (Log4jContextFactory.java:241)
>         at org.apache.logging.log4j.spi.LoggerContext org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(java.lang.String, java.lang.ClassLoader, java.lang.Object, boolean, java.net.URI, java.lang.String) (Log4jContextFactory.java:46)
>         at org.apache.logging.log4j.spi.LoggerContext org.apache.logging.log4j.LogManager.getContext(boolean) (LogManager.java:118)
>         at org.apache.logging.log4j.Logger org.apache.logging.log4j.LogManager.getLogger(java.lang.String) (LogManager.java:622)
>         at org.apache.logging.log4j.Logger org.apache.logging.log4j.LogManager.getRootLogger() (LogManager.java:657)
>         at void com.example.log4japi.MainActivity.<init>() (MainActivity.java:28)
>         at java.lang.Object java.lang.Class.newInstance() (Class.java:-2)
>         at android.app.Activity android.app.AppComponentFactory.instantiateActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (AppComponentFactory.java:95)
>         at android.app.Activity androidx.core.app.CoreComponentFactory.instantiateActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (CoreComponentFactory.java:45)
>         at android.app.Activity android.app.Instrumentation.newActivity(java.lang.ClassLoader, java.lang.String, android.content.Intent) (Instrumentation.java:1253)
>         at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:3353)
>         at android.app.Activity android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.app.servertransaction.PendingTransactionActions, android.content.Intent) (ActivityThread.java:3601)
>         at void android.app.servertransaction.LaunchActivityItem.execute(android.app.ClientTransactionHandler, android.os.IBinder, android.app.servertransaction.PendingTransactionActions) (LaunchActivityItem.java:85)
>         at void android.app.servertransaction.TransactionExecutor.executeCallbacks(android.app.servertransaction.ClientTransaction) (TransactionExecutor.java:135)
>         at void android.app.servertransaction.TransactionExecutor.execute(android.app.servertransaction.ClientTransaction) (TransactionExecutor.java:95)
>         at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:2066)
>         at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
> I/xample.log4jap:     at void android.os.Looper.loop() (Looper.java:223)
>         at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:7656)
>         at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
>         at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:592)
>         at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:947)

This seems to occur whenever an instance of org.apache.logging.log4j.Logger is returned (in the sample I will upload shortly, defined in the Activity scope). This does not cause any application crash however, and on further inspection in v2.24.1 I can see that once the logging configuration is set appropriately for the logger in use, logging commands are working as expected (pre-2.24 this was not the case!)

The issues I'm currently facing seem related to log4j2 config initialisation at this point. I only seem to be able to use the root logger: `Logger log = LogManager.getRootLogger();` 

which by default has LogLevel.ERROR set. This can be overridden using e.g. `Configurator.setLevel(log, Level.INFO);`

If I try to use any other logger: `Logger log = LogManager.getLogger(MainActivity.class);`

again, Level.ERROR is the default, but I am then unable to change the logging level. This is presumably because no configuration is available, so no other loggers are defined. It seems I can't get the application to read in a configuration from src/main/assets (which should be available on the log4j classpath).

**log4j2.properties:**

```
# Root logger configuration
status = error
name = PropertiesConfig

# Appenders
appender.console.type = Console
appender.console.name = Console
appender.console.target = SYSTEM_OUT
appender.console.layout.type = PatternLayout
appender.console.layout.pattern = [%d{yyyy-MM-dd HH:mm:ss.SSS}] %-5p- %m%n

# Root logger level and appenders
rootLogger.level = INFO
rootLogger.appenderRefs = stdout
rootLogger.appenderRef.stdout.ref = Console

logger.com.example.log4japi.MainActivity.name = com.example.log4japi.MainActivity
logger.com.example.log4japi.MainActivity.level = DEBUG
logger.com.example.log4japi.MainActivity.additivity = false
logger.com.example.log4japi.MainActivity.appenderRefs = stdout
logger.com.example.log4japi.MainActivity.appenderRef.stdout.ref = Console
```

Attempts to manually load using:

```
InputStream inputStream = getAssets().open("log4j2.properties");
ConfigurationSource source = new ConfigurationSource(inputStream);
context.start(ConfigurationFactory.getInstance().getConfiguration(context, source));
```

result in error:

>java.lang.UnsupportedOperationException: This parser does not support specification "Unknown" version "0.0"

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/config/xml/XmlConfiguration.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/config/xml/XmlConfiguration.java
@@ -216,14 +216,16 @@ public class XmlConfiguration extends AbstractConfiguration implements Reconfigu
      */
     private static void enableXInclude(final DocumentBuilderFactory factory) {
         try {
-            // Alternative: We set if a system property on the command line is set, for example:
-            // -DLog4j.XInclude=true
             factory.setXIncludeAware(true);
             // LOG4J2-3531: Xerces only checks if the feature is supported when creating a factory. To reproduce:
             // -Dorg.apache.xerces.xni.parser.XMLParserConfiguration=org.apache.xerces.parsers.XML11NonValidatingConfiguration
-            factory.newDocumentBuilder();
-        } catch (final UnsupportedOperationException | ParserConfigurationException e) {
-            factory.setXIncludeAware(false);
+            try {
+                factory.newDocumentBuilder();
+            } catch (final ParserConfigurationException e) {
+                factory.setXIncludeAware(false);
+                LOGGER.warn("The DocumentBuilderFactory [{}] does not support XInclude: {}", factory, e);
+            }
+        } catch (final UnsupportedOperationException e) {
             LOGGER.warn("The DocumentBuilderFactory [{}] does not support XInclude: {}", factory, e);
         } catch (final AbstractMethodError | NoSuchMethodError err) {
             LOGGER.warn(

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/impl/Log4jContextFactory.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/impl/Log4jContextFactory.java
@@ -29,6 +29,7 @@ import org.apache.logging.log4j.core.config.ConfigurationFactory;
 import org.apache.logging.log4j.core.config.ConfigurationSource;
 import org.apache.logging.log4j.core.config.DefaultConfiguration;
 import org.apache.logging.log4j.core.config.composite.CompositeConfiguration;
+import org.apache.logging.log4j.core.selector.BasicContextSelector;
 import org.apache.logging.log4j.core.selector.ClassLoaderContextSelector;
 import org.apache.logging.log4j.core.selector.ContextSelector;
 import org.apache.logging.log4j.core.util.Cancellable;
@@ -36,6 +37,7 @@ import org.apache.logging.log4j.core.util.Constants;
 import org.apache.logging.log4j.core.util.DefaultShutdownCallbackRegistry;
 import org.apache.logging.log4j.core.util.Loader;
 import org.apache.logging.log4j.core.util.ShutdownCallbackRegistry;
+import org.apache.logging.log4j.core.util.internal.SystemUtils;
 import org.apache.logging.log4j.spi.LoggerContextFactory;
 import org.apache.logging.log4j.status.StatusLogger;
 import org.apache.logging.log4j.util.PropertiesUtil;
@@ -104,7 +106,12 @@ public class Log4jContextFactory implements LoggerContextFactory, ShutdownCallba
         } catch (final Exception e) {
             LOGGER.error("Unable to create custom ContextSelector. Falling back to default.", e);
         }
-        return new ClassLoaderContextSelector();
+        // StackLocator is broken on Android:
+        // 1. Android could use the StackLocator implementation for JDK 11, but does not support multi-release JARs.
+        // 2. Android does not have the `sun.reflect` classes used in the JDK 8 implementation.
+        //
+        // Therefore, we use a single logger context.
+        return SystemUtils.isOsAndroid() ? new BasicContextSelector() : new ClassLoaderContextSelector();
     }
 
     private static ShutdownCallbackRegistry createShutdownCallbackRegistry() {

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/lookup/JmxRuntimeInputArgumentsLookup.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/lookup/JmxRuntimeInputArgumentsLookup.java
@@ -17,10 +17,13 @@
 package org.apache.logging.log4j.core.lookup;
 
 import java.lang.management.ManagementFactory;
-import java.util.List;
+import java.util.Collections;
 import java.util.Map;
+import org.apache.logging.log4j.Logger;
 import org.apache.logging.log4j.core.LogEvent;
 import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.util.internal.SystemUtils;
+import org.apache.logging.log4j.status.StatusLogger;
 
 /**
  * Maps JVM input arguments (but not main arguments) using JMX to acquire JVM arguments.
@@ -31,17 +34,16 @@ import org.apache.logging.log4j.core.config.plugins.Plugin;
 @Plugin(name = "jvmrunargs", category = StrLookup.CATEGORY)
 public class JmxRuntimeInputArgumentsLookup extends MapLookup {
 
-    static {
-        final List<String> argsList = ManagementFactory.getRuntimeMXBean().getInputArguments();
-        JMX_SINGLETON = new JmxRuntimeInputArgumentsLookup(MapLookup.toMap(argsList));
-    }
+    private static final Logger LOGGER = StatusLogger.getLogger();
 
-    public static final JmxRuntimeInputArgumentsLookup JMX_SINGLETON;
+    public static final JmxRuntimeInputArgumentsLookup JMX_SINGLETON = new JmxRuntimeInputArgumentsLookup();
 
     /**
      * Constructor when used directly as a plugin.
      */
-    public JmxRuntimeInputArgumentsLookup() {}
+    public JmxRuntimeInputArgumentsLookup() {
+        this(getMapFromJmx());
+    }
 
     public JmxRuntimeInputArgumentsLookup(final Map<String, String> map) {
         super(map);
@@ -54,5 +56,16 @@ public class JmxRuntimeInputArgumentsLookup extends MapLookup {
         }
         final Map<String, String> map = getMap();
         return map == null ? null : map.get(key);
+    }
+
+    private static Map<String, String> getMapFromJmx() {
+        if (!SystemUtils.isOsAndroid()) {
+            try {
+                return MapLookup.toMap(ManagementFactory.getRuntimeMXBean().getInputArguments());
+            } catch (LinkageError e) {
+                LOGGER.warn("Failed to get JMX arguments from JVM.", e);
+            }
+        }
+        return Collections.emptyMap();
     }
 }

--- a/log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/SystemUtils.java
+++ b/log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/SystemUtils.java
@@ -1,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.util.internal;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.status.StatusLogger;
+
+public final class SystemUtils {
+
+    private static final Logger LOGGER = StatusLogger.getLogger();
+
+    private static String getJavaVendor() {
+        try {
+            return System.getProperty("java.vendor");
+        } catch (final SecurityException e) {
+            LOGGER.warn("Unable to determine Java vendor.", e);
+        }
+        return "Unknown";
+    }
+
+    public static boolean isOsAndroid() {
+        return getJavaVendor().contains("Android");
+    }
+
+    private SystemUtils() {}
+}

--- a/src/changelog/.2.x.x/3056_android_support.xml
+++ b/src/changelog/.2.x.x/3056_android_support.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<entry xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns="https://logging.apache.org/xml/ns"
+       xsi:schemaLocation="https://logging.apache.org/xml/ns https://logging.apache.org/xml/ns/log4j-changelog-0.xsd"
+       type="fixed">
+  <issue id="3056" link="https://github.com/apache/logging-log4j2/issues/3056"/>
+  <description format="asciidoc">Fix Android-related issues in Log4j Core.</description>
+</entry>

--- a/src/site/antora/modules/ROOT/pages/faq.adoc
+++ b/src/site/antora/modules/ROOT/pages/faq.adoc
@@ -327,3 +327,55 @@ https://gradleup.com/shadow/[Gradle Shadow Plugin]:::
 You need to use the
 https://github.com/GradleUp/shadow/blob/main/src/main/groovy/com/github/jengelman/gradle/plugins/shadow/transformers/Log4j2PluginsCacheFileTransformer.groovy[`Log4j2PluginsCacheFileTransformer`].
 ====
+
+[#android]
+== Can I use Log4j with Android?
+
+Of course, you can!
+Since version `2.25.0` both the Log4j API and our three Log4j API implementations are tested for compatibility with the Android platform.
+
+If you use
+xref:manual/api.adoc[Log4j API]
+in an Android project, you have four choices for the Log4j API implementation:
+
+[#android-log4j-core]
+Log4j Core::
++
+Our
+xref:manual/implementation.adoc[reference Log4j API implementation]
+works on Android out-of-the-box.
+However, due to the limitations of the Android platform, the following features will not work:
++
+* The
+xref:manual/configuration.adoc#xinclude[XInclude feature]
+for XML configuration files will not work if you are using the standard Android XML parser.
+You might need to add the
+https://xerces.apache.org/[Xerces parser]
+to use the feature.
+* Due to the lack of Android support for multi-release JARs, some location-based features like the no-arg
+link:javadoc/log4j-api/org/apache/logging/log4j/LogManager.html#getLogger()[`LogManager.getLogger()`]
+method or
+xref:manual/systemproperties.adoc#log4j2.contextSelector[`ClassLoaderContextSelector`]
+(default on JRE) are not available.
+You should use `BasicContextSelector` (default on Android) or `BasicAsyncLoggerContextSelector` instead.
+
+[#android-jul]
+JUL::
+[#android-logback]
+Logback::
++
+Both our
+xref:manual/installation.adoc#impl-jul[Log4j API-to-JUL]
+and
+xref:manual/installation.adoc#impl-logback[Log4j API-to-SLF4J]
+bridges are tested for compatibility with Android.
+
+[#android-native]
+Log4j API-to-Android logging API bridge::
++
+If you wish to bridge Log4j API to
+https://developer.android.com/reference/android/util/Log[Android's native logging API]
+directly, you can use the **third-party** `com.celeral:log4j2-android` artifact.
+See the
+https://github.com/Celeral/log4j2-android[`log4j2-android` project website]
+for more information.

--- a/src/site/antora/modules/ROOT/partials/manual/systemproperties/properties-context-selector.adoc
+++ b/src/site/antora/modules/ROOT/partials/manual/systemproperties/properties-context-selector.adoc
@@ -26,7 +26,11 @@
 | link:../javadoc/log4j-core/org/apache/logging/log4j/core/selector/ContextSelector.html[`Class<? extends ContextSelector>`]
 
 | Default value
-| link:../javadoc/log4j-core/org/apache/logging/log4j/core/selector/ClassLoaderContextSelector.html[`ClassLoaderContextSelector`]
+|
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/selector/ClassLoaderContextSelector.html[`ClassLoaderContextSelector`]
+
+(on Android)
+link:../javadoc/log4j-core/org/apache/logging/log4j/core/selector/BasicContextSelector.html[`BasicContextSelector`]
 |===
 
 Specifies the fully qualified class name of the

